### PR TITLE
Refactor chunks and event between elements

### PR DESCRIPTION
**Pull request by jzaehrin**

### CODE REVIEW

File: src/forza/chunk.rs [hunk 1]
```diff
@@ -1,30 +1,160 @@
...
+
+pub type ChunkId = usize;
+pub type LapId = Option<u16>;
+
+#[derive(PartialEq, Default, Clone, Copy)]
+pub struct ChunkSelector(pub ChunkId, pub LapId);
+
+pub enum ChunksEvent {
+    LastChunk(ChunkSelector, GameMode),
+}
```
Comment (ttreyer) [inline src/forza/chunk.rs]: UI stuff should not go in `forza` lib.

File: src/forza/chunk.rs [hunk 2]
```diff
@@ -1,30 +1,160 @@
...
+    fn retrieve_events(&mut self) -> Option<HashMap<u8, Event>> {
+        if self.events.is_empty() {
+            return None;
+        }
 
-        chunks.back_mut().unwrap().push(p);
+        let events = replace(&mut self.events, HashMap::with_capacity(3));
+        Some(events)
     }
 }
```
Comment (ttreyer) [inline src/forza/chunk.rs]: Same as above. No UI stuff in here.

File: src/gui/chunk_panel.rs [hunk 3]
```diff
@@ -1,43 +1,67 @@
...
-        self.selection = ChunkSelection(chunk_id, lap_id)
+    pub fn new() -> Self {
+        Self {
+            selection: ChunkSelector::default(),
+            events: HashMap::with_capacity(1),
+        }
+    }
+
+    fn select(&mut self, chunk_id: ChunkId, lap_id: LapId) {
+        self.selection = ChunkSelector(chunk_id, lap_id);
```
Comment (ttreyer) [inline src/gui/chunk_panel.rs:31]: Is this still needed if the selection is passed within the event?

File: src/gui/control_panel.rs [hunk 4]
```diff
@@ -1,21 +1,35 @@
...
 use egui::{CtxRef, Ui};
 
-use crate::dialog;
+use crate::{
+    dialog,
+    event::{Event, EventGenerator},
+};
 
-pub enum ControlAction {
+#[repr(u8)]
```
Comment (ttreyer) [inline src/gui/control_panel.rs]:
```
type EventType = u8;
#[repr(EventType)]
```
Would that work?

File: src/gui/control_panel.rs [hunk 5]
```diff
@@ -1,21 +1,35 @@
...
+    events: HashMap<u8, Event>,
 }
 
 impl ControlPanel {
+    pub fn new() -> Self {
+        Self {
+            record: bool::default(),
+            next_race: bool::default(),
+            events: HashMap::with_capacity(3),
+        }
```
Comment (ttreyer) [inline src/gui/control_panel.rs]:
```
Self {
  events: HashMap::with_capacity(3),
  Default::defauts()
}
```

**Review by ttreyer — COMMENTED** [review]
Could we have the event map in `App`, passed down to the panels, instead of each panel having its own event generator?

File: src/gui/chunk_panel.rs (hunk 3 above)
Comment (jzaehrin) [reply src/gui/chunk_panel.rs:30]: Correct, it still not required

File: src/gui/control_panel.rs (hunk 4 above)
Comment (jzaehrin) [reply src/gui/control_panel.rs]: I think is not possible, only primitive type name is allowed. But anymore required due to the casting as u8.

File: src/gui/chunk_panel.rs (hunk 3 above)
Comment (jzaehrin) [reply src/gui/chunk_panel.rs:31]: It's already used to generate the Event and update self.selection.

**Pull request merged**